src/openrct2/peep/GuestPathfinding.h

```diff
@@ -37,19 +37,43 @@ extern RideId gPeepPathFindQueueRideIndex;
 // In practice, if this is false, gPeepPathFindQueueRideIndex is always RIDE_ID_NULL.
 extern bool gPeepPathFindIgnoreForeignQueues;
 
-// Given a peep 'peep' at tile 'loc', who is trying to get to 'gPeepPathFindGoalPosition', decide
-// the direction the peep should walk in from the current tile.
-Direction peep_pathfind_choose_direction(const TileCoordsXYZ& loc, Peep* peep);
+class GuestPathfinding
+{
+public:
+    virtual ~GuestPathfinding() = default;
 
-// Test whether the given tile can be walked onto, if the peep is currently at height currentZ and
-// moving in direction currentDirection.
-bool IsValidPathZAndDirection(TileElement* tileElement, int32_t currentZ, int32_t currentDirection);
+    // Given a peep 'peep' at tile 'loc', who is trying to get to 'gPeepPathFindGoalPosition', decide
+    // the direction the peep should walk in from the current tile.
+    virtual Direction peep_pathfind_choose_direction(const TileCoordsXYZ& loc, Peep* peep) = 0;
 
-// Overall guest pathfinding AI. Sets up Peep::DestinationX/DestinationY (which they move to in a
-// straight line, no pathfinding). Called whenever the guest has arrived at their previously set destination.
-//
-// Returns 0 if the guest has successfully had a new destination set up, nonzero otherwise.
-int32_t guest_path_finding(Guest* peep);
+    // Test whether the given tile can be walked onto, if the peep is currently at height currentZ and
+    // moving in direction currentDirection.
+    static bool IsValidPathZAndDirection(TileElement* tileElement, int32_t currentZ, int32_t currentDirection);
+
+    // Overall guest pathfinding AI. Sets up Peep::DestinationX/DestinationY (which they move to in a
+    // straight line, no pathfinding). Called whenever the guest has arrived at their previously set destination.
+    //
+    // Returns 0 if the guest has successfully had a new destination set up, nonzero otherwise.
+    virtual int32_t guest_path_finding(Guest* peep) = 0;
+};
+
+class OriginalPathfinding : public GuestPathfinding
+{
+public:
+    Direction peep_pathfind_choose_direction(const TileCoordsXYZ& loc, Peep* peep) final override;
+
+    int32_t guest_path_finding(Guest* peep) final override;
+
+private:
+    int32_t GuestPathFindParkEntranceEntering(Peep* peep, uint8_t edges);
+
+    int32_t GuestPathFindPeepSpawn(Peep* peep, uint8_t edges);
+
+    int32_t GuestPathFindParkEntranceLeaving(Peep* peep, uint8_t edges);
+};
+
+// TODO: Implement a better solution than a global variable for the utilized pathfinder
+extern GuestPathfinding* gGuestPathfinder;
```

Comment:
What kind of better solution were you thinking of?

Reply:
I don't know really tbh, this code is so reliant on global variables though, it would probably be a lot of work to do anything else.

Reply:
Your recommendation to change that to a unique_ptr solved the majority of my concerns with leaving that as a global.

Reply:
Just becareful as you can not control the order of global unique_ptr destruction. So sometimes globals get disposed before other globals still use them during their dispose code.